protocol/binary/protocol.go

```diff
@@ -161,54 +166,117 @@ func (p *Protocol) DecodeRequest(et wire.EnvelopeType, r io.ReaderAt) (wire.Valu
 	return val, NoEnvelopeResponder, err
 }
 
-// noEnvelopeResponder responds to a request without an envelope.
-type noEnvelopeResponder struct{}
+// ReadRequest reads off the request envelope (if present) from an io.Reader,
+// populating the provided BodyReader to read off the full request struct,
+// asserting the EnvelopeType (either OneWay or Unary) if an envlope exists.
+// A ResponseWriter that understands the enveloping used is returned.
+//
+// This allows a Thrift request handler to transparently read requests
+// regardless of whether the caller is configured to submit envelopes.
+//
+// This is possible because we can distinguish an envelope from a bare request
+// struct by looking at the first byte and the length of the message.
+//
+// 1. A message of length 1 containing only 0x00 can only be an empty struct.
+// 0x00 is the type ID for STOP, indicating the end of the struct.
+//
+// 2. A message of length >1 starting with 0x00 can only be a non-strict
+// envelope (not versioned), assuming the message name is less than 16MB long.
+// In this case, the first four bytes indicate the length of the method name,
+// which is unlikely to overflow into the high byte.
+//
+// 3. A message of length >1, where the first byte is <0 can only be a strict envelope.
+// The MSB indicates that the message is versioned. Reading the first two bytes
+// and masking out the MSB indicates the version number.
+// At this time, there is only one version.
+//
+// 4. A message of length >1, where the first byte is >=0 can only be a bare
+// struct starting with that field identifier. Valid field identifiers today
+// are in the range 0x00-0x0f. There is some chance that a future version of
+// the protocol will add more field types, but it is very unlikely that the
+// field type will flow into the MSB (128 type identifiers, starting with the
+// 15 valid types today).
+func (p *Protocol) ReadRequest(
+	ctx context.Context,
+	et wire.EnvelopeType,
+	r io.Reader,
+	body stream.BodyReader,
+) (stream.ResponseWriter, error) {
+	var buf [2]byte
 
-func (noEnvelopeResponder) EncodeResponse(v wire.Value, t wire.EnvelopeType, w io.Writer) error {
-	return Default.Encode(v, w)
-}
+	sr := p.Reader(bytes.NewReader(buf[:]))
+	if count, _ := r.Read(buf[0:2]); count < 2 {
+		return NoEnvelopeResponder, body.Decode(sr)
```

Comment:
This bit is a bit off. We don't need to instantiate sr until we get inside the if block, and at then, we should still close the reader so that it gets returned to the pool.

Separately, `buf[:]` isn't right. If we read only one of the two bytes, that's all we're going to want to parse. I'll fix this bit.

Reply:
Fixed with https://github.com/thriftrw/thriftrw-go/pull/526/commits/6fc9df673c9f6564b0c25280e8a9dc8b9de29c61